### PR TITLE
CHROMEOS Add LAVA template for QEMU tast tests

### PR DESCRIPTION
**Pull request by nuclearcat**
As QEMU version of ChromiumOS have slightly different boot
sequence and uses tast built from ChromiumOS sources, it require separate LAVA template.

Signed-off-by: Denys Fedoryshchenko <denys.f@collabora.com>

### CODE REVIEW

**Comment by nuclearcat** [comment]
I will retest it first

File: config/lava/chromeos/chromiumosvm-tast-template.jinja2 [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    media: tmpfs
+#    docker:
+#      image: kernelci/qemu
+#      binary: qemu-system-x86_64
+    timeout:
+      minutes: 5
+    prompts:
+    - "login:"
+
+- test:
```
Comment (gctucker) [inline config/lava/chromeos/chromiumosvm-tast-template.jinja2:35]: I thought this part would be moved to a `tast.jinja` template and shared between QEMU and Chromebook tests?

Comment (nuclearcat) [reply config/lava/chromeos/chromiumosvm-tast-template.jinja2:35]: Will do.

Comment (nuclearcat) [reply config/lava/chromeos/chromiumosvm-tast-template.jinja2]: Done

**Comment by nuclearcat** [comment]
Generated template manually and run: https://lava.collabora.co.uk/scheduler/job/5858857
Still need to add custom kernel support, but i think it will be done, after we add building "chromeos" kernel, where tast tests are functional.

**Comment by gctucker** [comment]
> Still need to add custom kernel support, but i think it will be done, after we add building "chromeos" kernel, where tast tests are functional.

I think we'll add support for running KernelCI tests with generic kernels (mainline, stable, next etc.) even if the tests are not working. Then some kernel fixes will be required to make the tests pass.  It's separate from building Chrome OS kernels.

**Comment by nuclearcat** [comment]
This is latest template (but needs https://github.com/kernelci/kernelci-core/pull/1074 ) with docker of kernelci
https://lava.collabora.co.uk/scheduler/job/5859493
This is same test, working properly i used updated docker image:
https://lava.collabora.co.uk/scheduler/job/5859529

File: config/core/test-configs.yaml [hunk 2]
```diff
@@ -2411,6 +2436,10 @@ test_configs:
     test_plans:
       - baseline_qemu
 
+  - device_type: qemu_x86_64-uefi-chromiumos
+    test_plans:
+      - chromiumosvm-tast
```
Comment (gctucker) [inline config/core/test-configs.yaml:2441]: Might be worth adding baseline_qemu, as a reference point if Chrome OS jobs fail to boot for some reason.

Comment (nuclearcat) [reply config/core/test-configs.yaml:2441]: Will do

File: config/core/test-configs.yaml [hunk 3]
```diff
@@ -1376,6 +1383,24 @@ device_types:
...
+    <<: *qemu_x86_64
+    context:
+      arch: x86_64
+      cpu: 'qemu64'
+      guestfs_interface: 'virtio'
+      extra_options:
+        - '-device pcie-root-port,port=0x10,chassis=1,id=pci.1,bus=pcie.0,multifunction=on,addr=0x2'
+        - '-device pcie-root-port,port=0x15,chassis=6,id=pci.6,bus=pcie.0,addr=0x2.0x5'
+        - '-device qemu-xhci,p2=15,p3=15,id=usb,bus=pci.6,addr=0x0'
+        - -device usb-storage,bus=usb.0,drive=lavatest2
```
Comment (gctucker) [inline config/core/test-configs.yaml:1396]: No quotes for this parameter?

File: config/lava/chromeos/tast.jinja2 [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      - repository:
+          metadata:
+            format: Lava-Test Test Definition 1.0
+            name: docker-ping
+          run:
+            steps:
+              - 'cd /home/cros-tast'
+{%- if cros_tast_tarball %}
+              - 'curl -s {{cros_tast_tarball}} | tar xjf -'
+{% endif %}
```
Comment (gctucker) [inline config/lava/chromeos/tast.jinja2:22]: Is it not mandatory to have a Tast tarball to run this job?

File: config/lava/chromeos/tast.jinja2 [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+{%- if rootfs_url %}
+              - 'curl -s {{ rootfs_url }}remote_test_runner -o remote_test_runner;chmod +x remote_test_runner'
+              - 'curl -s {{ rootfs_url }}tast -o tast;chmod +x tast'
+              - 'cp remote_test_runner /usr/bin/remote_test_runner'
+{% endif %}
```
Comment (gctucker) [inline config/lava/chromeos/tast.jinja2:27]: The `rootfs_url` parameter is used unconditionally in the `deploy` action, so it looks like it's mandatory too.

Comment (gctucker) [reply config/lava/chromeos/tast.jinja2:27]: Maybe we should leave `rootfs_url` as the actual URL for the rootfs image, and have other template parameters to download the Tast stuff.

File: config/lava/chromeos/tast.jinja2 (hunk 4 above)
Comment (nuclearcat) [reply config/lava/chromeos/tast.jinja2:22]: We might have either rootfs_url in case of ChromiumOS opensource build (then we fetch relevant tast and remote_test_runner), or we might have cros_tast_tarball, tast tarball supplied locally (as we had on proprietary build tests)

File: config/core/test-configs.yaml (hunk 3 above)
Comment (nuclearcat) [reply config/core/test-configs.yaml]: Fixed that

File: config/lava/chromeos/tast.jinja2 (hunk 5 above)
Comment (nuclearcat) [reply config/lava/chromeos/tast.jinja2:27]: Just tested now and it works as designed, here is diff my test to @alexandrasp test on real hardware:
`
 - test:
       timeout:
@@ -101,9 +85,7 @@
           run:
             steps:
               - 'cd /home/cros-tast'
-              - 'curl -s http://storage.chromeos.kernelci.org/images/rootfs/chromeos/chromiumos-amd64-generic/crostest003/remote_test_runner -o remote_test_runner;chmod +x remote_test_runner'
-              - 'curl -s http://storage.chromeos.kernelci.org/images/rootfs/chromeos/chromiumos-amd64-generic/crostest003/tast -o tast;chmod +x tast'
-              - 'cp remote_test_runner /usr/bin/remote_test_runner'
+              - 'curl -s http://172.17.0.1/chromeos/autotest_server_package_octopus_R92_13981.0.0.tar.bz2 | tar xjf -'
 
               - 'while ! ping -c 1 -w 1 $(lava-target-ip); do sleep 1; done'
               - 'ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -i /home/cros-tast/.ssh/id_rsa root@$(lava-target-ip) whoami'


`

File: config/lava/chromeos/chromiumosvm-tast-template.jinja2 [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        image_arg: -bios {bios}
+        url: http://storage.kernelci.org/images/uefi/edk2-stable202005/x86_64/OVMF.fd
+      rootfs:
+        image_arg: -drive format=raw,file={rootfs},id=lavatest2,if=none
+        url: {{ rootfs_url }}chromiumos_test_image.bin
+        compression: gz
+{%- if kernel_url %}
+#       kernel:
+#         image_arg: -kernel {kernel} -append "init=/sbin/init boot=local rootwait ro noresume noswap loglevel=7 earlyprintk=serial,keep console=tty0 console=ttyS0 i915.modeset=1 cros_efi cros_debug root=/dev/sda3"
+#         url: {{ kernel_url }}
```
Comment (gctucker) [inline config/lava/chromeos/chromiumosvm-tast-template.jinja2:22]: We probably should resolve this somehow.  I would suggest:
* get some Chromium OS tests on QEMU scheduled by KernelCI
* get the results on chromeos.kernelci.org web dashboard to confirm it's all working
* un-comment this part to enable the KernelCI kernels to be used (linux-next) and see if some config works with QEMU
* if the KenrelCI kernels don't work then add a patch to comment out this section, so we can drop it later

Comment (nuclearcat) [reply config/lava/chromeos/chromiumosvm-tast-template.jinja2:22]: I will add tast tarball generation in builder script.

Comment (gctucker) [reply config/lava/chromeos/chromiumosvm-tast-template.jinja2:22]: If it's failing all the tests and they're reported as failures then that's good enough for KernelCI. But I think we should go through each step mentioned above unless you want to suggest something else.

Comment (gctucker) [reply config/lava/chromeos/chromiumosvm-tast-template.jinja2:22]: About extra patches, that's something to discuss with Google folks and others. Ideally we shouldn't be applying anything as KernelCI is to test upstream, and report all the things that aren't working.

Comment (nuclearcat) [reply config/lava/chromeos/chromiumosvm-tast-template.jinja2]: I uncommented it and i will schedule tests after also i change build script to generate tast tgz

File: config/lava/chromeos/chromiumosvm-tast-template.jinja2 [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    timeout:
+      minutes: {{ job_timeout }}
+    to: tmpfs
+
+- boot:
+    method: qemu
+    media: tmpfs
+#    docker:
+#      image: kernelci/qemu
+#      binary: qemu-system-x86_64
```
[the remark this answers is not in the document]
Comment (nuclearcat) [reply config/lava/chromeos/chromiumosvm-tast-template.jinja2:34]: Done

File: config/lava/chromeos/tast.jinja2 (hunk 4 above)
Comment (nuclearcat) [reply config/lava/chromeos/tast.jinja2:22]: Done

File: config/lava/chromeos/chromiumosvm-tast-template.jinja2 (hunk 6 above)
Comment (gctucker) [reply config/lava/chromeos/chromiumosvm-tast-template.jinja2]: But how about getting results in KernelCI first before trying with real kernel builds?

I'm trying to help get this resolved hence the list of steps I'm suggesting but I don't understand the approach of trying with KernelCI build now as we haven't proved we can get results with a working kernel all the way to the web dashboard. I don't think the jobs are even submitted yet, so that's something to check and fix first.

Comment (nuclearcat) [reply config/lava/chromeos/chromiumosvm-tast-template.jinja2]: I mean i will try to build rootfs image with tast tarball changes first, then i will generate LAVA definition manually (by kci_test) with upstream kernel and will check how it goes and what results we get on LAVA.

**Comment by gctucker** [comment]
@nuclearcat I've rebased the `chromeos` branch, please rebase this PR.

**Comment by nuclearcat** [comment]
Done

**Comment by gctucker** [comment]
Here's one problem:
```
$ ./kci_test submit --lab-config=lab-collabora --jobs=data/jobs/*qemu*
ERROR data/jobs/next-master-next-20220318-x86_64-x86_64_defconfig+x86-chromebook-clang-13-no-dtb-qemu_x86_64-uefi-chromiumos-chromiumosvm-tast.yaml: while parsing a block mapping
  in "<unicode string>", line 61, column 7
did not find expected key
  in "<unicode string>", line 68, column 8
```
It seems to also happen with the `octopus` Tast test definition:
```
ERROR data/jobs/next-master-next-20220318-x86_64-x86_64_defconfig+x86-chromebook-clang-13-no-dtb-hp-x360-12b-n4000-octopus-chromeos-tast.yaml: while parsing a block mapping
  in "<unicode string>", line 2, column 1
did not find expected key
  in "<unicode string>", line 71, column 1
```
```

**Comment by gctucker** [comment]
To reproduce this:
```
cd kernelci-core
git remote update origin
git checkout origin/chromeos.kernelci.org
./kci_test get_lab_info --lab-config=lab-collabora
mkdir -p data/chromeos-x86
rm -f data/chromeos/x86/*
wget -P data/chromeos-x86 https://storage.chromeos.kernelci.org/next/master/next-20220318/x86_64/x86_64_defconfig%2Bx86-chromebook/clang-13/{steps,bmeta,artifacts}.json
mkdir -p data/jobs
rm -f data/jobs/*
./kci_test generate --lab-config=lab-collabora --install-path=data/chromeos-x86 --output=data/jobs
./kci_test submit --lab-config=lab-collabora --jobs=data/jobs/*qemu*
./kci_test submit --lab-config=lab-collabora --jobs=data/jobs/*octopus*
```

**Comment by gctucker** [comment]
Here's the problem, indentation is not consistent in the template file:
```diff
diff --git a/config/lava/chromeos/chromiumosvm-tast-template.jinja2 b/config/lava/chromeos/chromiumosvm-tast-template.jinja2
index 7aba6c59de32..da8b0860eacf 100644
--- a/config/lava/chromeos/chromiumosvm-tast-template.jinja2
+++ b/config/lava/chromeos/chromiumosvm-tast-template.jinja2
@@ -17,9 +17,9 @@ actions:
         url: {{ rootfs_url }}chromiumos_test_image.bin
         compression: gz
 {%- if kernel_url %}
-       kernel:
-         image_arg: -kernel {kernel} -append "init=/sbin/init boot=local rootwait ro noresume noswap loglevel=7 earlyprintk=serial,keep console=tty0 console=ttyS0 i915.modeset=1 cros_efi cros_debug root=/dev/sda3"
-         url: {{ kernel_url }}
+      kernel:
+        image_arg: -kernel {kernel} -append "init=/sbin/init boot=local rootwait ro noresume noswap loglevel=7 earlyprintk=serial,keep console=tty0 console=ttyS0 i915.modeset=1 cros_efi cros_debug root=/dev/sda3"
+        url: {{ kernel_url }}
 {%- endif %}
     os: oe
     timeout:
```

**Comment by gctucker** [comment]
To generate _working_ job definitions with the Chrome OS builds:
```
./kci_test generate --lab-config=lab-collabora --install-path=data/chromeos-x86 --output=data/jobs --db-config=staging.kernelci.org --storage=https://storage.chromeos.kernelci.org/
```

**Comment by nuclearcat** [comment]
There is several changes i did, related to compressing rootfs image, making tast as tgz and etc, and just updated relevant parts.
We have still missing major bit - kernel modules need to be somehow installed into qemu disk image.

**Comment by nuclearcat** [comment]
Without custom kernel part it works: https://lava.collabora.co.uk/scheduler/job/5930429
Now i have to figure out how to deploy modules.

File: config/lava/chromeos/tast.jinja2 (hunk 4 above)
Comment (gctucker) [reply config/lava/chromeos/tast.jinja2:22]: If it's mandatory then we should remove the `{%- if` statement.

File: config/lava/chromeos/tast.jinja2 [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+          run:
+            steps:
+              - 'cd /home/cros-tast'
+{%- if cros_tast_tarball %}
+              - 'curl -s {{cros_tast_tarball}} | tar xjf -'
+{% endif %}
+{%- if rootfs_url %}
+              - 'curl -s {{rootfs_url}}tast.tgz | tar xzf -'
+              - 'cp remote_test_runner /usr/bin/remote_test_runner'
+{% endif %}
```
Comment (gctucker) [inline config/lava/chromeos/tast.jinja2:26]: And if we're always using the tarball URL then we should be able to drop this too I believe.

**Review by gctucker — COMMENTED** [review]
Next step is to get this run by KernelCI to have results on chromeos.kernelci.org with the built-in kernel from the QEMU image. We can look into adding modules and a kernel image as a follow-up (and in fact, it's the same as with any other QEMU job with modules and a kernel image over TFTP).

**Comment by gctucker** [comment]
The last jobs on staging still had the gz rootfs issue:
https://lava.collabora.co.uk/scheduler/job/5925889

Starting another job now to check whether the latest version of this PR is working as expected.

**Comment by gctucker** [comment]
It looks like one job got scheduled in LAVA, but it seemed to get stuck in a reboot loop:
https://lava.collabora.co.uk/scheduler/job/5930819

File: config/lava/chromeos/chromiumosvm-tast-template.jinja2 [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        url: http://storage.kernelci.org/images/uefi/edk2-stable202005/x86_64/OVMF.fd
+      rootfs:
+        image_arg: -drive format=raw,file={rootfs},id=lavatest2,if=none
+        url: {{ rootfs_url }}chromiumos_test_image.bin.gz
+        compression: gz
+{%- if kernel_url %}
+      kernel:
+        image_arg: -kernel {kernel} -append "init=/sbin/init boot=local rootwait ro noresume noswap loglevel=7 earlyprintk=serial,keep console=tty0 console=ttyS0 i915.modeset=1 cros_efi cros_debug root=/dev/sda3"
+        url: {{ kernel_url }}
+{%- endif %}
```
Comment (gctucker) [inline config/lava/chromeos/chromiumosvm-tast-template.jinja2:23]: We only just got one QEMU job running in KernelCI, and it's not booting with linux-next (and no modules installed yet). So I think this part should be left out for now to keep using the built-in kernel instead until we have the rest sorted out.

Comment (nuclearcat) [reply config/lava/chromeos/chromiumosvm-tast-template.jinja2]: Done, removed it.

File: config/lava/chromeos/tast.jinja2 [hunk 10]
```diff
@@ -0,0 +1,32 @@
+- test:
+      timeout:
+        minutes: 45
+      docker:
+        image: {{ docker_image }}
+        volumes:
+          - /tmp/octopus:/home/cros-tast/octopus
```
[the remark this answers is not in the document]
Comment (nuclearcat) [reply config/lava/chromeos/tast.jinja2:7]: I will test without it now

[the remark this answers is not in the document]
Comment (nuclearcat) [reply config/lava/chromeos/tast.jinja2]: Done, works without this lines https://lava.collabora.co.uk/scheduler/job/5932725
I pushed updated commit

**Comment by gctucker** [comment]
@nuclearcat Please rebase

**Comment by nuclearcat** [comment]
Done

File: config/lava/chromeos/tast.jinja2 (hunk 8 above)
Comment (nuclearcat) [reply config/lava/chromeos/tast.jinja2:24]: Sent PR https://github.com/kernelci/kernelci-core/pull/1083

**Comment by nuclearcat** [comment]
Added fixed_kernel parameter and tested generation manually.
Also during next build it was scheduled and tested by jenkins:
https://chromeos.kernelci.org/test/plan/id/6243c175d628922963b58d29/

File: config/core/test-configs.yaml [hunk 11]
```diff
@@ -1382,6 +1394,31 @@ device_types:
...
+        - '-device qemu-xhci,p2=15,p3=15,id=usb,bus=pci.6,addr=0x0'
+        - '-device usb-storage,bus=usb.0,drive=lavatest2'
+        - '-m 2048'
+        - '-device usb-tablet,id=input0,bus=usb.0,port=2'
+        - '-device pcie-root-port,port=0x11,chassis=2,id=pci.2,bus=pcie.0,addr=0x2.0x1'
+        - '-device virtio-vga,id=video0,virgl=on,max_outputs=1,bus=pcie.0,addr=0x3'
+        - '-device ich9-intel-hda,id=sound0,bus=pcie.0,addr=0x1b'
+        - '-machine pc-q35-5.2,accel=kvm,usb=off,vmport=off,dump-guest-core=off'
+    filters:
+      - passlist: {defconfig: ['x86-chromebook', 'defconfig']}
```
Comment (gctucker) [inline config/core/test-configs.yaml:1420]: It would be good to understand why some jobs aren't being generated with no filters at all. In theory, there shouldn't be a need for adding these to a passlist filter.

File: config/lava/chromeos/chromeos-tast-template.jinja2 [hunk 12]
```diff
@@ -5,44 +5,22 @@
 {% endblock %}
 {% block actions %}
 actions:
-  - deploy:
+- deploy:
```
Comment (gctucker) [inline config/lava/chromeos/chromeos-tast-template.jinja2:8]: I think the indentation was fine, what's the reason for this change?

Comment (nuclearcat) [reply config/lava/chromeos/chromeos-tast-template.jinja2:8]: Jenkins job was failing.
Here is failed example: https://bot.staging.kernelci.org/job/chromeos/job/test-runner/5182/console
I generated yaml manually also and confirmed that octopus was failing.

File: config/lava/chromeos/tast.jinja2 [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+        image: {{ docker_image }}
+        wait:
+          device: true
+      results:
+          location: /home/cros-tast/lava
+      definitions:
+      - repository:
+          metadata:
+            format: Lava-Test Test Definition 1.0
+            name: docker-ping
```
Comment (gctucker) [inline config/lava/chromeos/tast.jinja2:14]: Wrong name

File: config/lava/chromeos/tast.jinja2 [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+{%- if rootfs_url %}
+              - 'curl -s {{rootfs_url}}tast.tgz | tar xzf -'
+              - 'cp remote_test_runner /usr/bin/remote_test_runner'
+{% endif %}
+              - 'while ! ping -c 1 -w 1 $(lava-target-ip); do sleep 1; done'
+              - 'ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -i /home/cros-tast/.ssh/id_rsa root@$(lava-target-ip) whoami'
+              - './tast-parser.py example.ARCFixture example.ManyParams.arc_state example.Param.dog example.Pass example.Fail'
+        from: inline
+        name: docker-tast-tests
+        path: inline/docker-ping.yaml
```
Comment (gctucker) [inline config/lava/chromeos/tast.jinja2:30]: Ditto

**Review by gctucker — COMMENTED** [review]
Some minor thing to tidy up.

About QEMU in Docker, we can't at the moment have both QEMU in a container and Tast running in a container on the host with network access from Tast to QEMU. This will require a bit more work, so we'll stick to using QEMU natively on the dispatcher for now.

File: config/lava/chromeos/tast.jinja2 (hunk 14 above)
Comment (nuclearcat) [reply config/lava/chromeos/tast.jinja2]: Fixed and tested in https://lava.collabora.co.uk/scheduler/job/5987415

File: config/lava/chromeos/tast.jinja2 (hunk 13 above)
Comment (nuclearcat) [reply config/lava/chromeos/tast.jinja2]: Fixed and tested in https://lava.collabora.co.uk/scheduler/job/5987415
